common/src/main/java/io/netty/util/ResourceLeakDetector.java

```diff
@@ -461,11 +461,19 @@ public boolean close(T trackedObject) {
             // Ensure that the object that was tracked is the same as the one that was passed to close(...).
             assert trackedHash == System.identityHashCode(trackedObject);
 
-            // We need to actually do the null check of the trackedObject after we close the leak because otherwise
-            // we may get false-positives reported by the ResourceLeakDetector. This can happen as the JIT / GC may
-            // be able to figure out that we do not need the trackedObject anymore and so already enqueue it for
-            // collection before we actually get a chance to close the enclosing ResourceLeak.
-            return close() && trackedObject != null;
+            // Recent versions of the JDK have a nasty habit of prematurely deciding objects are unreachable.
+            // see: https://stackoverflow.com/questions/26642153/finalize-called-on-strongly-reachable-object-in-java-8
+            // The method Reference.reachabilityFence offers a solution to this problem.
+            // However, besides only being available in Java 9+,
+            // it "is designed for use in uncommon situations of premature finalization where using
+            // synchronized blocks or methods [is] not possible or do not provide the desired control."
+            // Because somewhere up the callstack should be a (successful) `trackedObject.release`,
+            // it is unreasonable that anyone else, anywhere, is hold a lock on the trackedObject.
+            // Therefore, it seems using a synchronization block is possible here, so we will use one!
+            synchronized (trackedObject)
+            {
```

Comment:
can you move the `{` up one line to match our coding style ?

Reply:
Would it be more clear if written this way?
```
boolean b = close();
synchronized (trackedObject) {}
return b;
```

I know that looks weird, but it makes it clear we aren't relying on the synchronization when running the close() method.

Reply:
@ejona86 I wonder if it could just optimise it away when it detects that the block is empty ?

Reply:
Nope, because it still has side-effects. See my comment in https://github.com/grpc/grpc-java/pull/1526/files#r55582622